src/sharechain/in_memory.rs

```diff
@@ -0,0 +1,214 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use log::{debug, info, warn};
+use minotari_app_grpc::tari_rpc::{NewBlockCoinbase, SubmitBlockRequest};
+use tari_common_types::tari_address::TariAddress;
+use tari_core::blocks::BlockHeader;
+use tari_utilities::epoch_time::EpochTime;
+use tari_utilities::hex::Hex;
+use tokio::sync::{RwLock, RwLockWriteGuard};
+
+use crate::sharechain::error::{BlockConvertError, Error};
+use crate::sharechain::{Block, ShareChain, ShareChainResult, MAX_BLOCKS_COUNT, SHARE_COUNT};
+
+const LOG_TARGET: &str = "in_memory_share_chain";
+
+pub struct InMemoryShareChain {
+    max_blocks_count: usize,
+    blocks: Arc<RwLock<Vec<Block>>>,
+}
+
+impl Default for InMemoryShareChain {
+    fn default() -> Self {
+        Self {
+            max_blocks_count: MAX_BLOCKS_COUNT,
+            blocks: Arc::new(RwLock::new(vec![
+                // genesis block
+                Block::builder().with_height(0).build(),
+            ])),
+        }
+    }
+}
+
+#[allow(dead_code)]
+impl InMemoryShareChain {
+    pub fn new(max_blocks_count: usize) -> Self {
+        Self {
+            max_blocks_count,
+            blocks: Arc::new(RwLock::new(vec![
+                // genesis block
+                Block::builder().with_height(0).build(),
+            ])),
+        }
+    }
+
+    async fn miners_with_shares(&self) -> HashMap<String, f64> {
+        let mut result: HashMap<String, f64> = HashMap::new(); // target wallet address -> number of shares
+        let blocks_read_lock = self.blocks.read().await;
+        blocks_read_lock.iter().for_each(|block| {
+            if let Some(miner_wallet_address) = block.miner_wallet_address() {
+                let addr = miner_wallet_address.to_base58();
+                if let Some(curr_hash_rate) = result.get(&addr) {
+                    result.insert(addr, curr_hash_rate + 1.0);
+                } else {
+                    result.insert(addr, 1.0);
+                }
+            }
+        });
+
+        result
+    }
+
+    async fn validate_block(&self, last_block: &Block, block: &Block) -> ShareChainResult<bool> {
+        // check if we have this block as last
+        if last_block == block {
+            warn!(target: LOG_TARGET, "↩️ This block already added, skip");
+            return Ok(false);
+        }
+
+        // validate hash
+        if block.hash() != block.generate_hash() {
+            warn!(target: LOG_TARGET, "❌ Invalid block, hashes do not match");
+            return Ok(false);
+        }
+
+        // validate height
+        if last_block.height() + 1 != block.height() {
+            warn!(target: LOG_TARGET, "❌ Invalid block, invalid block height: {:?} != {:?}", last_block.height() + 1, block.height());
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+
+    async fn submit_block_with_lock(
+        &self,
+        blocks: &mut RwLockWriteGuard<'_, Vec<Block>>,
+        block: &Block,
+        clear_before_add: bool,
```

Comment:
what is the meaning of this parameter, based on the name?

i see this method is used twice:
1. to submit one block through `ShareChain::submit_block` with `false` value; and
2. to submit multiple blocks through `ShareChain::submit_blocks`, used by `Service::handle_share_chain_sync_response`, with `true` value.

is the latter case blocks that are not stable / potentially being re-orged or something?

Reply:
Yeah the variable name can be a bit misleading, so the thing is that when we synchronize blocks from another peer (as we are newcomers to p2pool for example) there is a big chance that the smallest height of the share chain is not 0 or 1, so we cannot validate because lowest height is more than the available blocks, so in this case we just clear all the blocks and receive everything. If it is set to false, then we try to apply new incoming blocks to the current one we have.